lib/src/frontend/expect.dart

```diff
@@ -68,38 +91,55 @@ void expect(actual, matcher,
     }
 
     Invoker.current.skip(message);
-    return;
+    return new Future.value();
+  }
+
+  if (matcher is AsyncMatcher) {
+    // Avoid async/await so that expect() throws synchronously when possible.
+    var result = matcher.matchAsync(actual);
+    expect(result, anyOf([
+      equals(null),
+      new isInstanceOf<Future>(),
+      new isInstanceOf<String>()
+    ]), reason: "matchAsync() may only return a String, a Future, or null.");
+
+    if (result is String) {
+      fail(formatFailure(matcher, actual, result, reason: reason));
+    } else if (result is Future) {
+      Invoker.current.addOutstandingCallback();
+      return result.then((realResult) {
+        if (realResult == null) return;
+        fail(formatFailure(matcher, actual, realResult, reason: reason));
+      }).whenComplete(() {
+        // Always remove this, in case the failure is caught and handled
+        // gracefully.
+        Invoker.current.removeOutstandingCallback();
+      });
+    }
+
+    return new Future.value();
   }
 
   var matchState = {};
   try {
-    if (matcher.matches(actual, matchState)) return;
+    if (matcher.matches(actual, matchState)) return new Future.value();
   } catch (e, trace) {
-    if (reason == null) {
-      reason = '${(e is String) ? e : e.toString()} at $trace';
-    }
+    reason ??= '$e at $trace';
   }
-  if (formatter == null) formatter = _defaultFailFormatter;
   fail(formatter(actual, matcher, reason, matchState, verbose));
+  return new Future.value();
 }
 
 /// Convenience method for throwing a new [TestFailure] with the provided
 /// [message].
 void fail(String message) => throw new TestFailure(message);
 
 // The default error formatter.
-String _defaultFailFormatter(
-    actual, Matcher matcher, String reason, Map matchState, bool verbose) {
-  var description = new StringDescription();
-  description.add('Expected: ').addDescriptionOf(matcher).add('\n');
-  description.add('  Actual: ').addDescriptionOf(actual).add('\n');
-
-  var mismatchDescription = new StringDescription();
-  matcher.describeMismatch(actual, mismatchDescription, matchState, verbose);
-
-  if (mismatchDescription.length > 0) {
-    description.add('   Which: ${mismatchDescription}\n');
-  }
-  if (reason != null) description.add(reason).add('\n');
-  return description.toString();
+String formatFailure(Matcher expected, actual, String which, {String reason}) {
+  var buffer = new StringBuffer();
+  buffer.writeln(indent(prettyPrint(expected),       first: 'Expected: '));
+  buffer.writeln(indent(prettyPrint(actual),         first: '  Actual: '));
+  if (which.isNotEmpty) buffer.writeln(indent(which, first: '   Which: '));
+  if (reason != null) buffer.writeln(reason);
```

Comment:
No, the extra whitespace is just to visually verify that the `first` prefixes are all the same width.